Firestore/core/src/firebase/firestore/timestamp.cc

```diff
@@ -29,46 +29,73 @@
 
 namespace firebase {
 
-Timestamp::Timestamp() {
+namespace {
+
+// We pass it by value without std::move.
+static_assert(std::is_trivially_copyable<Timestamp>::value,
+              "Timestamp must be trivially copyable");
+
+constexpr int32_t kNanosPerSecond = 1E9;
+
+/**
+ * Creates a `Timestamp` from the given non-normalized inputs.
+ *
+ * Timestamp protos require `Timestamp` to always has a positive number of
+ * nanoseconds that is counting forward. For negative time, we need to adjust
+ * representations with negative nanoseconds. That is make (negative seconds s1
+ * + negative nanoseconds ns1) to (negative seconds s2 + positive nanoseconds
+ * ns2). Since nanosecond part is always less than 1 second in our
+ * representation, instead of starting at s1 and going back ns1 nanoseconds,
+ * start at (s1 minus one second) and go *forward* ns2 = (1 second + ns1, ns1 <
+ * 0) nanoseconds.
+ */
+Timestamp MakeNormalizedTimestamp(int64_t seconds, int64_t nanos) {
```

Comment:
Very optional: overall, I like this refactoring. The only nitpick I have is that this function takes plain integers, whereas it would have been a little more type-aware if it took `chr::seconds` and `chr::nanoseconds`. OTOH, if you find that to be additional boilerplate for no gain, then just leave as is.

Reply:
The problem is that only conversion from `std::time_point` naturally gets values of this type and that otherwise the values aren't naturally of this type. I find `std::chrono` to be byzantine and never remember how to do even the simplest of tasks with it. All told I'm much more comfortable pushing that to the edges of the system and otherwise use simpler things internally.

If anything I'd prefer to remove more of our `std::chrono` usage and use `absl::Time` internally.